### PR TITLE
Create of secrets (sshpublickey) from file

### PR DESCRIPTION
**Pull request by mad01**
MVP implementation of adding secret from file related to #2195 
Note: most of the code is from `make apimachinery`

example 
```bash
cat > secret.yaml <<EOF
apiVersion: kops/v1alpha2
kind: Secret
metadata:
  labels:
    kops.k8s.io/cluster: dev.k8s.example.com
spec:
  name: "admin"
  sshPublicKey: ""
EOF
```

`kops create -f secret.yaml`

### CODE REVIEW

**Comment by mad01** [comment]
I think this is good to go

**Comment by justinsb** [comment]
Good job figuring out the apimachinery stuff :-)

I uploaded a PR where I also started working on this (I was working more on the kops-server side), so we have something to compare it against; the API object I had created is here: https://github.com/kubernetes/kops/pull/3150/files#diff-eae71bc4c338f07057b68a92f56c78eaR23

So a few points of comparison:

* Because k8s has a type called "Secret", I don't think we want to call it "Secret".  I called it "KeySet" because...
* The KeySet has a notion of different versions, because particularly when it comes to the PKI stuff we might have several certificates that are recognized (although I think we'll only ever have a single active private key).  
* The SSH public key is bit of a special case.  It isn't really managed very much by kops, and we just pass it to the cloud.  Maybe that means it should have its own API type (but then I would worry that our use is so small that we shouldn't be doing that)
* The third usage is for tokens, which are gradually going away but are still used.  KeySet has a Type field which differentiates between the 3 types.

Rather than having a Type and then parsing the values, you've gone more explicit (though you've only handled the sshPublicKey case).  On reflection, I think I like your approach there, it feels user friendly.  We would then add another field `secretToken` or something for a secret token, and `certificates` and `privateKey` fields for PKI.  We would use validation to ensure that you don't set both.  That to me feels more user-friendly.

So TLDR, I like your approach - going through the code now :-)

File: docs/secrets.md [hunk 1]
```diff
@@ -24,3 +24,18 @@ The ID form can be used when there are multiple matching keys.
...
+
+Secret spec
+```
+apiVersion: kops/v1alpha2
+kind: Secret
+metadata:
+  labels:
+    kops.k8s.io/cluster: dev.k8s.example.com
+spec:
+  name: "admin"
```
Comment (justinsb) [inline docs/secrets.md]: metadata already has a name field, so I think we should be using that name field.

File: pkg/apis/kops/secret.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+package kops
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient=true
+
+// Secret represent a set of kops secrets
+type Secret struct {
```
Comment (justinsb) [inline pkg/apis/kops/secret.go]: I don't think we should call it Secret just for confusion (I think they are namespaced by the group, so it is technically permitted, but all the same it is confusing).  Perhaps `SystemSecret` or `SecretMaterial` or `ClusterSecret` or something like that...  I did `KeySet` before which I think is fine but weaker than the 3 I just mentioned.  WDYT?

File: pkg/apis/kops/secret.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec SecretSpec `json:"spec,omitempty"`
+}
+
+type SecretSpec struct {
+	SshPublicKey string `json:"sshPublicKey,omitempty"`
+	Name         string `json:"name,omitempty"`
```
Comment (justinsb) [inline pkg/apis/kops/secret.go]: I think we should use the name in the metadata.

**Comment by justinsb** [comment]
Proposed renaming the type to something other than Secret, and not having a second Name field.  Other than that, LooksGreatToMe.

/ok-to-test

**Comment by mad01** [comment]
@justinsb makes since will take a look at changing name `Secret` since that is overlapping the k8s version of it. Makes since it can be confusing to have the same name when they have nothing related

**Comment by justinsb** [comment]
Awesome @mad01 - thanks!  Can you just ping me when it's ready for review, as otherwise it's easy to miss on github :-)

**Comment by mad01** [comment]
closing since something went wrong
